### DISCUSSION TITLE
Loading all children given a parent ID

### DISCUSSION THREAD

**Discussion by sflahave** [graphql-java/graphql-java#2552] [Q&A] [answered]
Suppose you're using Hibernate/JPA, and you have a one-to-many relationship. For example, Order and LineItem, where Order has-many LineItems, and the LineItems are loaded lazily. 

Suppose you have an `OrderResolver`, which has a `lineItems` method. You want to set up a DataLoader to load all LineItems given an Order ID.  How would you do this with DataLoader? 
I.e., you know the Order ID but you don't know the line item IDs because you haven't loaded them yet, so you can't pass a list of LineItem IDs. Any suggestions?

**Comment by lqc** [comment]
Well it depends on your system, use-case, size of the data, etc.

For some cases, just reading the LineItem ids first and passing them to the DataLoader might be the most efficient way to do it. In other cases, you might need to do that because you want the LineItems filtered by things that are not in your DB, so you need to fetch the ids from an ElasticSearch (or similar) first and load their details later.

If you assume the simplest case, then you don't need a DataLoader at all. Just let the OrderResolver fetch the collection eagerly and return it.

If for some reason you still need a DataLoader, you can have load all the items in ``lineItems`` using a plain JPA query and then use ``DataLoader.prime()`` to put the LineItems into the DataLoader's cache. Or you could disable DataLoader's cache at all and just rely on Hibernate by using ``Session.byMultipleIds`` in the DataLoader.

**Comment by bbakerman** [comment, marked as answer]
DataLoaders primary purpose is that of a "batch loader" - it has some caching but other systems do caching better.

DataLoader is there to get "child" objects by key in batch.

So using your example, DataLoader would work best if you had "5 Orders" and you want to LineItems for them in match.  So instead of calling "load line items" 5 times, a DataLoader allows you to build them and batch them

You might then load LineItems for all 4 all at once.

This only works if you have the "child keys" of course - DataLoader is a system to back load a value from a key - `K -> V`

**Comment by sflahave** [comment]
Thanks for the explanations, @bbakerman and @lqc !
